assembly/src/assembler/mod.rs

```diff
@@ -376,8 +375,11 @@ pub fn combine_blocks(mut blocks: Vec<CodeBlock>) -> CodeBlock {
         }
     }
 
-    debug_assert!(!blocks.is_empty(), "no blocks");
-    blocks.remove(0)
+    if blocks.is_empty() {
+        CodeBlock::new_span(vec![Operation::Noop])
+    } else {
+        blocks.remove(0)
+    }
```

Comment:
A couple comments here:

First, I wonder if it would be cleaner to do this check outside of `combine_blocks()` (i.e., on line 287 above). It could look something like:
```Rust
if blocks.is_empty() {
    Ok(CodeBlock::new_span(vec![Operation::Noop]))
} else {
    Ok(combine_blocks(blocks))
}
```

Second, I wonder if instead of creating a span block with a `Noop`, we should create an empty span block (i.e., `CodeBlock::new_span()`. We will need to update the definition of `Span` block to allow empty blocks and also update the processor to make sure it handles empty span blocks correctly - but it might be a cleaner approach and could make special handling of such blocks easier in the future (i.e. we could add something like `Span::is_empty()`).

Reply:
About your first point:

At the moment, `combine_blocks` takes a `Vec<CodeBlock>` as an argument.  In general, it's useful for functions to cover their entire domain as far as possible.  In this case, we can give a reasonable answer for an empty vector.

However, I do see what you mean!  If you want to write functions that don't accept empty inputs, I would suggest letting the compiler (and the programmer) now, for example by changing the argument type to `NonEmpty<CodeBlock>` via the Crate [nonempty](https://docs.rs/nonempty/latest/nonempty/).

About your second point:

What you say makes sense!  Though I wonder whether we should put that in a future PR?  I even wonder whether an empty `span` should be the canonical representation here?  I mean in principle we could also change the system to support empty `join` blocks.

Of course, I'm not saying the latter is a good idea, just saying that it's not completely clear that the empty `span` is the only conceivable option.

Apropos options: another possible fix is to make `compile_body` return `Result<Option<CodeBlock>, AssemblyError>` instead, where `None` means essentially `noop`, and make all the other changes to support this style.  Including perhaps moving to `NonEmpty` where appropriate?

But that's for another PR..

Reply:
> for example by changing the argument type to `NonEmpty<CodeBlock>` via the Crate [nonempty](https://docs.rs/nonempty/latest/nonempty/).

I agree, this would be the right way to do it, but our general philosophy is to try to minimize the number of external dependencies as much as possible (at least for non-test code). So, if it was a part of Rust's standard library, we'd totally use it. But given the situation, I think using a clear assert at the beginning of the function is an acceptable alternative.

> What you say makes sense! Though I wonder whether we should put that in a future PR?

I guess this depends on how complex the change is. If it is just changing things in a few places, would be good to include it in this PR so that we can close this. But if it is an involved change, we can certainly create an issue and come beck to it in future PRs. My thinking is that is probably the former, but I also didn't look into this too deeply - so, could be wrong.

Reply:
@bobbinth I can apply your suggestion of
```rust
if blocks.is_empty() {
    Ok(CodeBlock::new_span(vec![Operation::Noop]))
} else {
    Ok(combine_blocks(blocks))
}
```
if that's what's needed to merge?

Reply:
Yes - that would be great!

Reply:
I moved the check out of `combine_blocks`.

Reply:
Looks good! But what is the reason for doing `if let [] = blocks[..]` instead of a simple `blocks.is_empty()`?

Reply:
I guess that's my Haskell background showing: no big reason, in my mind general pattern matching is just simpler than generating a Boolean and then using a special purpose matching construct that only works on them (if-then-else). There's also always a small chance of accidentally flipping the meaning of your booleans (eg by overlooking a hidden negation) and the compiler can't help you at all.

In this case, there's really no difference and likely produces the same optimised native code.

I can change it, if you like?

P.S. For background reasoning on my style choice see https://web.archive.org/web/20190518151856/http://blakesmith.me/2019/05/07/rust-patterns-enums-instead-of-booleans.html

 Or https://runtimeverification.com/blog/code-smell-boolean-blindness/ or https://existentialtype.wordpress.com/2011/03/15/boolean-blindness/

Reply:
Makes sense! And I like the reasoning - but I would still prefer to change it, if you don't mind. The motivation is that we don't use this approach anywhere else, and I'd rather keep the codebase as consistent as possible.

Reply:
Sure, will change.